### ISSUE TITLE
[go_router_builder] Generate a method to return the matching `GoRoute` or `GoRouteData` from a location

### ISSUE THREAD

**Issue by ValentinVignal** [flutter/flutter#107993]
## Use case

Let's consider an app with 2 routes:

```dart
// A route to log in when the user is logged out.
@TypedGoRoute<LoginRoute>(path: '/login')
class LoginRoute extends GoRouteData {
  const LoginRoute();

  @override
  Widget build(BuildContext context) => const LoginScreen();
}

// The home route the user sees when logged in.
@TypedGoRoute<HomeRoute>(path: '/home')
class HomeRoute extends GoRouteData {
  const HomeRoute();

  @override
  Widget build(BuildContext context) => const HomeScreen();
}

final router = GoRouter(
  routes: $appRoutes,
  redirect: (GoRouterState state) {
    // Here we only have access to the state which contains strings (`location`).
  
    if(userIsLoggedIn()) {
      // Verify the route is not the login route.
    } else {
      // Verify the route is not the home route.
    }
  },
);
```

This generates the extensions:

```dart
extension $LoginRouteExtension on LoginRoute {
  static LoginRoute _fromState(GoRouterState state) => const LoginRoute();

  String get location => GoRouteData.$location(
        '/login',
      );

  void go(BuildContext context) => context.go(location, extra: this);

  void push(BuildContext context) => context.push(location, extra: this);
}

// Home extension
// ...
```

There is no way to get the route from the URL/location (including nested ones) and this forces us to do some logic on the URL/location string which I think defeats the purpose of `go_router_builder`.

The `route` is actually accessible with `state.extra`. But it is available only if the code use `push`/`go` of the generated extension.
This is not the case when the user enters the URL by hand in the address bar/clicks on the back or forward button of the browser/refreshes the page.
Also, because of this issue (https://github.com/flutter/flutter/issues/106790), it is not possible to use the generated `go` and `push` method from the extensions with an abstracted code.

In short, we cannot rely on `state.extra` to get the route.

## Proposal

It would be very useful to generate a method that parses an URL/location and returns the correct route.

```dart
final router = GoRouter(
  routes: $appRoutes,
  redirect: (GoRouterState state) {
    final route = $someMethod$(state.location);
    // Typed logic 
  },
);
```

**Comment by btrautmann** [comment]
Definitely agree with the proposal, and I'd add one more hope/wish (that I'm not sure is possible), but having a way to refer back to the original `GoRouteData` via a mapping of `GoRoute` to `GoRouteData` or otherwise would be helpful.

For example, consider wanting to be able to label routes as requiring the user to be authenticated or not:

```dart
class MyRoute extends GoRouteData with AuthenticationRequired {
  @override
  Widget build(BuildContext context) {
    return const MyRouteScreen();
  }
}
```

With the above, we'd be able to do something like:

```dart
// `Uri` has been intercepted via native embedding code and is
// now being processed before proceeding
...
final path = uri.path;
final route = $someMethod$(path);
final routeData = $routeData[route];
if (routeData is AuthenticationRequired) {
  // Send user to login
} else {
  // User can proceed without authentication
}
```
